File: tests/wasm/test_mfi.js
/**
 * WASM binding tests for MFI (Money Flow Index) indicator.
 * These tests mirror the Rust unit tests to ensure WASM bindings work correctly.
 */
import test from 'node:test';
import assert from 'node:assert';
import path from 'path';
import { fileURLToPath } from 'url';
import { 
    loadTestData, 
    assertArrayClose, 
    assertClose,
    isNaN,
    assertAllNaN,
    assertNoNaN,
    EXPECTED_OUTPUTS
} from './test_utils.js';
import { compareWithRust } from './rust-comparison.js';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

let wasm;
let testData;

test.before(async () => {
    
    try {
        const wasmPath = path.join(__dirname, '../../pkg/my_project.js');
        const importPath = process.platform === 'win32' 
            ? 'file:///' + wasmPath.replace(/\\/g, '/')
            : wasmPath;
        wasm = await import(importPath);
        
    } catch (error) {
        console.error('Failed to load WASM module. Run "wasm-pack build --features wasm --target nodejs" first');
        throw error;
    }
    
    testData = loadTestData();
});

test('MFI partial params', () => {
    
    const high = new Float64Array(testData.high);
    const low = new Float64Array(testData.low);
    const close = new Float64Array(testData.close);
    const volume = new Float64Array(testData.volume);
    
    
    const typicalPrice = new Float64Array(high.length);
    for (let i = 0; i < high.length; i++) {
        typicalPrice[i] = (high[i] + low[i] + close[i]) / 3.0;
    }
    
    const result = wasm.mfi_js(typicalPrice, volume, 14);
    assert.strictEqual(result.length, typicalPrice.length);
});

test('MFI accuracy', async () => {
    
    const high = new Float64Array(testData.high);
    const low = new Float64Array(testData.low);
    const close = new Float64Array(testData.close);
    const volume = new Float64Array(testData.volume);
    const expected = EXPECTED_OUTPUTS.mfi;
    
    
    const typicalPrice = new Float64Array(high.length);
    for (let i = 0; i < high.length; i++) {
        typicalPrice[i] = (high[i] + low[i] + close[i]) / 3.0;
    }
    
    const result = wasm.mfi_js(
        typicalPrice,
        volume,
        expected.defaultParams.period
    );
    
    assert.strictEqual(result.length, typicalPrice.length);
    
    
    const last5 = result.slice(-5);
    assertArrayClose(
        last5,
        expected.last5Values,
        1e-1,  
        "MFI last 5 values mismatch"
    );
    
    
    await compareWithRust('mfi', result, 'hlc3_volume', expected.defaultParams);
});

test('MFI default candles', () => {
    
    const high = new Float64Array(testData.high);
    const low = new Float64Array(testData.low);
    const close = new Float64Array(testData.close);
    const volume = new Float64Array(testData.volume);
    
    
    const typicalPrice = new Float64Array(high.length);
    for (let i = 0; i < high.length; i++) {
        typicalPrice[i] = (high[i] + low[i] + close[i]) / 3.0;
    }
    
    const result = wasm.mfi_js(typicalPrice, volume, 14);
    assert.strictEqual(result.length, typicalPrice.length);
});

test('MFI zero period', () => {
    
    const data = new Float64Array(20).fill(100);
    const volume = new Float64Array(20).fill(1000);
    
    assert.throws(() => {
        wasm.mfi_js(data, volume, 0);
    }, /Invalid period/);
});

test('MFI period exceeds length', () => {
    
    const input_high = [1.0, 2.0, 3.0];
    const input_low = [0.5, 1.5, 2.5];
    const input_close = [0.8, 1.8, 2.8];
    const input_volume = [100.0, 200.0, 300.0];
    const typicalPrice = new Float64Array(3);
    for (let i = 0; i < 3; i++) {
        typicalPrice[i] = (input_high[i] + input_low[i] + input_close[i]) / 3.0;
    }
    const volume = new Float64Array(input_volume);
    
    assert.throws(() => {
        wasm.mfi_js(typicalPrice, volume, 10);
    }, /Invalid period/);
});

test('MFI very small dataset', () => {
    
    const typicalPrice = new Float64Array([(1.0 + 0.5 + 0.8) / 3.0]);
    const volume = new Float64Array([100.0]);
    
    assert.throws(() => {
        wasm.mfi_js(typicalPrice, volume, 14);
    }, /Invalid period/);
});

test('MFI empty input', () => {
    
    const empty = new Float64Array([]);
    
    assert.throws(() => {
        wasm.mfi_js(empty, empty, 14);
    }, /empty/i);
});

test('MFI mismatched array lengths', () => {
    
    const typicalPrice = new Float64Array([1, 2, 3]);
    const volume = new Float64Array([1, 2]);
    
    assert.throws(() => {
        wasm.mfi_js(typicalPrice, volume, 14);
    }, /empty/i);
});

test('MFI NaN handling', () => {
    
    const high = new Float64Array(testData.high);
    const low = new Float64Array(testData.low);
    const close = new Float64Array(testData.close);
    const volume = new Float64Array(testData.volume);
    
    
    const typicalPrice = new Float64Array(high.length);
    for (let i = 0; i < high.length; i++) {
        typicalPrice[i] = (high[i] + low[i] + close[i]) / 3.0;
    }
    
    const result = wasm.mfi_js(typicalPrice, volume, 14);
    assert.strictEqual(result.length, typicalPrice.length);
    
    
    assertAllNaN(result.slice(0, 13), "Expected NaN in warmup period");
    
    
    if (result.length > 13) {
        
        let hasValidValues = false;
        for (let i = 13; i < Math.min(100, result.length); i++) {
            if (!isNaN(result[i])) {
                hasValidValues = true;
                
                assert(result[i] >= 0.0 && result[i] <= 100.0,
                    `MFI value ${result[i]} at index ${i} out of bounds [0, 100]`);
            }
        }
        assert(hasValidValues, "Expected valid values after warmup period");
    }
});

test('MFI with NaN input', () => {
    
    const n = 30;
    const typicalPrice = new Float64Array(n);
    const volume = new Float64Array(n).fill(1000);
    for (let i = 0; i < n; i++) {
        typicalPrice[i] = i;
    }
    typicalPrice[5] = NaN; 

    const period = 14;
    const result = wasm.mfi_js(typicalPrice, volume, period);
    assert.strictEqual(result.length, n);

    
    
    
    for (let i = 0; i < period - 1; i++) {
        assert(isNaN(result[i]), `Expected NaN at warmup index ${i}`);
    }
    for (let i = period - 1; i < n; i++) {
        assert(isNaN(result[i]), `Expected NaN at index ${i} due to seed NaN`);
    }
});

test('MFI all NaN input', () => {
    
    const data = new Float64Array(20).fill(NaN);
    const volume = new Float64Array(20).fill(NaN);
    
    assert.throws(() => {
        wasm.mfi_js(data, volume, 14);
    }, /All values are NaN/);
});

test('MFI zero volume', () => {
    
    const n = 30;
    const typicalPrice = new Float64Array(n);
    const volume = new Float64Array(n).fill(0);
    
    
    for (let i = 0; i < n; i++) {
        typicalPrice[i] = 100 + i;
    }
    
    const result = wasm.mfi_js(typicalPrice, volume, 14);
    
    
    for (let i = 13; i < result.length; i++) {
        assert(Math.abs(result[i] - 0.0) < 1e-10,
            `Expected MFI to be 0 with zero volume, got ${result[i]} at index ${i}`);
    }
});

test('MFI batch single parameter set', () => {
    
    const high = new Float64Array(testData.high);
    const low = new Float64Array(testData.low);
    const close = new Float64Array(testData.close);
    const volume = new Float64Array(testData.volume);
    
    
    const typicalPrice = new Float64Array(high.length);
    for (let i = 0; i < high.length; i++) {
        typicalPrice[i] = (high[i] + low[i] + close[i]) / 3.0;
    }
    
    
    const config = {
        period_range: [14, 14, 0]  
    };
    
    const batchResult = wasm.mfi_batch(typicalPrice, volume, config);
    
    
    assert(batchResult.values, "Expected values in batch result");
    assert(batchResult.combos, "Expected combos in batch result");
    assert.strictEqual(batchResult.rows, 1, "Expected 1 row");
    assert.strictEqual(batchResult.cols, typicalPrice.length, "Expected cols to match data length");
    assert.strictEqual(batchResult.combos.length, 1, "Expected 1 parameter combination");
    
    
    const singleResult = wasm.mfi_js(typicalPrice, volume, 14);
    const firstRow = batchResult.values.slice(0, batchResult.cols);
    assertArrayClose(
        firstRow,
        singleResult,
        1e-10,
        "Batch should match single calculation"
    );
});

test('MFI batch multiple periods', () => {
    
    const high = new Float64Array(testData.high.slice(0, 100)); 
    const low = new Float64Array(testData.low.slice(0, 100));
    const close = new Float64Array(testData.close.slice(0, 100));
    const volume = new Float64Array(testData.volume.slice(0, 100));
    
    
    const typicalPrice = new Float64Array(100);
    for (let i = 0; i < 100; i++) {
        typicalPrice[i] = (high[i] + low[i] + close[i]) / 3.0;
    }
    
    
    const config = {
        period_range: [10, 20, 5]
    };
    
    const result = wasm.mfi_batch(typicalPrice, volume, config);
    
    
    assert.strictEqual(result.rows, 3, "Expected 3 rows (periods)");
    assert.strictEqual(result.cols, 100, "Expected 100 columns");
    assert.strictEqual(result.combos.length, 3, "Expected 3 combinations");
    assert.strictEqual(result.values.length, 300, "Expected 300 total values");
    
    
    const periods = [10, 15, 20];
    for (let i = 0; i < periods.length; i++) {
        const rowStart = i * 100;
        const rowEnd = rowStart + 100;
        const rowData = result.values.slice(rowStart, rowEnd);
        
        const singleResult = wasm.mfi_js(typicalPrice, volume, periods[i]);
        assertArrayClose(
            rowData,
            singleResult,
            1e-10,
            `Period ${periods[i]} mismatch`
        );
        
        
        for (let j = 0; j < periods[i] - 1; j++) {
            assert(isNaN(rowData[j]), `Expected NaN at warmup index ${j} for period ${periods[i]}`);
        }
    }
});

test('MFI batch metadata', () => {
    
    const typicalPrice = new Float64Array(50).fill(100);
    const volume = new Float64Array(50).fill(1000);
    
    const result = wasm.mfi_batch(typicalPrice, volume, {
        period_range: [10, 20, 5]  
    });
    
    
    assert.strictEqual(result.combos.length, 3);
    
    
    assert.strictEqual(result.combos[0].period, 10);
    assert.strictEqual(result.combos[1].period, 15);
    assert.strictEqual(result.combos[2].period, 20);
});

test('MFI batch edge cases', () => {
    
    const typicalPrice = new Float64Array([1, 2, 3, 4, 5, 6, 7, 8, 9, 10]);
    const volume = new Float64Array(10).fill(100);
    
    
    const singleBatch = wasm.mfi_batch(typicalPrice, volume, {
        period_range: [5, 5, 1]
    });
    
    assert.strictEqual(singleBatch.values.length, 10);
    assert.strictEqual(singleBatch.combos.length, 1);
    
    
    const largeBatch = wasm.mfi_batch(typicalPrice, volume, {
        period_range: [5, 7, 10]  
    });
    
    
    assert.strictEqual(largeBatch.values.length, 10);
    assert.strictEqual(largeBatch.combos.length, 1);
    
    
    assert.throws(() => {
        wasm.mfi_batch(new Float64Array([]), new Float64Array([]), {
            period_range: [14, 14, 0]
        });
    }, /All values are NaN/);
});

test('MFI memory allocation/deallocation', () => {
    
    const len = 1000;
    
    
    const ptr = wasm.mfi_alloc(len);
    assert(ptr !== 0, "Expected non-null pointer from mfi_alloc");
    
    
    wasm.mfi_free(ptr, len);
    
    
    wasm.mfi_free(0, len); 
});

test('MFI fast API (in-place)', { skip: 'WASM memory not directly accessible in ES modules' }, () => {
    
    const high = new Float64Array(testData.high.slice(0, 100));
    const low = new Float64Array(testData.low.slice(0, 100));
    const close = new Float64Array(testData.close.slice(0, 100));
    const volume = new Float64Array(testData.volume.slice(0, 100));
    
    
    const typicalPrice = new Float64Array(100);
    for (let i = 0; i < 100; i++) {
        typicalPrice[i] = (high[i] + low[i] + close[i]) / 3.0;
    }
    
    
    const expected = wasm.mfi_js(typicalPrice, volume, 14);
    
    
    const len = typicalPrice.length;
    const tpPtr = wasm.mfi_alloc(len);
    const volPtr = wasm.mfi_alloc(len);
    const outPtr = wasm.mfi_alloc(len);
    
    try {
        
        const tpMem = new Float64Array(wasm.memory.buffer, tpPtr, len);
        const volMem = new Float64Array(wasm.memory.buffer, volPtr, len);
        tpMem.set(typicalPrice);
        volMem.set(volume);
        
        
        wasm.mfi_into(
            tpPtr,
            volPtr,
            outPtr,
            len,
            14
        );
        
        
        const result = new Float64Array(wasm.memory.buffer, outPtr, len);
        
        
        assertArrayClose(
            Array.from(result),
            Array.from(expected),
            1e-10,
            "Fast API result mismatch"
        );
    } finally {
        wasm.mfi_free(tpPtr, len);
        wasm.mfi_free(volPtr, len);
        wasm.mfi_free(outPtr, len);
    }
});

test('MFI fast API with aliasing', { skip: 'WASM memory not directly accessible in ES modules' }, () => {
    
    const high = new Float64Array(testData.high.slice(0, 100));
    const low = new Float64Array(testData.low.slice(0, 100));
    const close = new Float64Array(testData.close.slice(0, 100));
    const volume = new Float64Array(testData.volume.slice(0, 100));
    
    
    const typicalPrice = new Float64Array(100);
    for (let i = 0; i < 100; i++) {
        typicalPrice[i] = (high[i] + low[i] + close[i]) / 3.0;
    }
    const typicalPriceCopy = new Float64Array(typicalPrice);
    
    
    const expected = wasm.mfi_js(typicalPrice, volume, 14);
    
    
    const len = 100;
    const tpPtr = wasm.mfi_alloc(len);
    const volPtr = wasm.mfi_alloc(len);
    
    try {
        
        const tpMem = new Float64Array(wasm.memory.buffer, tpPtr, len);
        const volMem = new Float64Array(wasm.memory.buffer, volPtr, len);
        tpMem.set(typicalPriceCopy);
        volMem.set(volume);
        
        
        wasm.mfi_into(
            tpPtr,
            volPtr,
            tpPtr,  
            len,
            14
        );
        
        
        const result = new Float64Array(wasm.memory.buffer, tpPtr, len);
        
        
        assertArrayClose(
            Array.from(result),
            Array.from(expected),
            1e-10,
            "Fast API aliasing result mismatch"
        );
    } finally {
        wasm.mfi_free(tpPtr, len);
        wasm.mfi_free(volPtr, len);
    }
});

test('MFI batch fast API', { skip: 'WASM memory not directly accessible in ES modules' }, () => {
    
    const high = new Float64Array(testData.high.slice(0, 50));
    const low = new Float64Array(testData.low.slice(0, 50));
    const close = new Float64Array(testData.close.slice(0, 50));
    const volume = new Float64Array(testData.volume.slice(0, 50));
    
    
    const typicalPrice = new Float64Array(50);
    for (let i = 0; i < 50; i++) {
        typicalPrice[i] = (high[i] + low[i] + close[i]) / 3.0;
    }
    
    
    const periods = { start: 10, end: 20, step: 5 }; 
    const numCombos = 3;
    const totalSize = numCombos * 50;
    
    
    const tpPtr = wasm.mfi_alloc(50);
    const volPtr = wasm.mfi_alloc(50);
    const outPtr = wasm.mfi_alloc(totalSize);
    
    try {
        
        const tpMem = new Float64Array(wasm.memory.buffer, tpPtr, 50);
        const volMem = new Float64Array(wasm.memory.buffer, volPtr, 50);
        tpMem.set(typicalPrice);
        volMem.set(volume);
        
        
        const rows = wasm.mfi_batch_into(
            tpPtr,
            volPtr,
            outPtr,
            50,
            periods.start,
            periods.end,
            periods.step
        );
        
        assert.strictEqual(rows, 3, "Expected 3 rows returned");
        
        
        const result = new Float64Array(wasm.memory.buffer, outPtr, totalSize);
        
        
        const singleResult = wasm.mfi_js(typicalPrice, volume, 10);
        const firstRow = Array.from(result.slice(0, 50));
        assertArrayClose(
            firstRow,
            Array.from(singleResult),
            1e-10,
            "Batch fast API first row mismatch"
        );
    } finally {
        wasm.mfi_free(tpPtr, 50);
        wasm.mfi_free(volPtr, 50);
        wasm.mfi_free(outPtr, totalSize);
    }
});

test('MFI zero-copy error handling', () => {
    
    assert.throws(() => {
        wasm.mfi_into(0, 0, 0, 10, 14);
    }, /null pointer|Null pointer/i);
    
    
    const ptr = wasm.mfi_alloc(10);
    const volPtr = wasm.mfi_alloc(10);
    try {
        
        assert.throws(() => {
            wasm.mfi_into(ptr, volPtr, ptr, 10, 0);
        }, /Invalid period/);
        
        
        assert.throws(() => {
            wasm.mfi_into(ptr, volPtr, ptr, 10, 20);
        }, /Invalid period/);
    } finally {
        wasm.mfi_free(ptr, 10);
        wasm.mfi_free(volPtr, 10);
    }
});

test('MFI memory leak prevention', { skip: 'WASM memory not directly accessible in ES modules' }, () => {
    
    const sizes = [100, 1000, 10000];
    
    for (const size of sizes) {
        const ptr = wasm.mfi_alloc(size);
        assert(ptr !== 0, `Failed to allocate ${size} elements`);
        
        
        const memView = new Float64Array(wasm.memory.buffer, ptr, size);
        for (let i = 0; i < Math.min(10, size); i++) {
            memView[i] = i * 1.5;
        }
        
        
        for (let i = 0; i < Math.min(10, size); i++) {
            assert.strictEqual(memView[i], i * 1.5, `Memory corruption at index ${i}`);
        }
        
        
        wasm.mfi_free(ptr, size);
    }
});

test.after(() => {
    console.log('MFI WASM tests completed');
});

File: tests/python/test_mfi.py
"""
Python binding tests for MFI (Money Flow Index) indicator.
These tests mirror the Rust unit tests to ensure Python bindings work correctly.
"""
import pytest
import numpy as np
import sys
from pathlib import Path

try:
    import my_project as ta_indicators
except ImportError:
    
    try:
        import my_project as ta_indicators
    except ImportError:
        pytest.skip("Python module not built. Run 'maturin develop --features python' first", allow_module_level=True)

from test_utils import load_test_data, assert_close, EXPECTED_OUTPUTS
from rust_comparison import compare_with_rust


class TestMfi:
    @pytest.fixture(scope='class')
    def test_data(self):
        return load_test_data()
    
    def test_mfi_partial_params(self, test_data):
        """Test MFI with default parameters - mirrors check_mfi_partial_params"""
        high = test_data['high']
        low = test_data['low']
        close = test_data['close']
        volume = test_data['volume']
        typical_price = (high + low + close) / 3.0
        
        
        result = ta_indicators.mfi(typical_price, volume, period=14)
        assert len(result) == len(typical_price)
    
    def test_mfi_accuracy(self, test_data):
        """Test MFI matches expected values from Rust tests - mirrors check_mfi_accuracy"""
        high = test_data['high']
        low = test_data['low']
        close = test_data['close']
        volume = test_data['volume']
        typical_price = (high + low + close) / 3.0
        expected = EXPECTED_OUTPUTS['mfi']
        
        result = ta_indicators.mfi(
            typical_price,
            volume,
            period=expected['default_params']['period']
        )
        
        assert len(result) == len(typical_price)
        
        
        assert_close(
            result[-5:],
            expected['last_5_values'],
            rtol=0.0,
            atol=1e-1,
            msg="MFI last 5 values mismatch"
        )
        
        
        compare_with_rust('mfi', result, 'hlc3_volume', expected['default_params'])
    
    def test_mfi_default_candles(self, test_data):
        """Test MFI with default parameters - mirrors check_mfi_default_candles"""
        high = test_data['high']
        low = test_data['low']
        close = test_data['close']
        volume = test_data['volume']
        typical_price = (high + low + close) / 3.0
        
        result = ta_indicators.mfi(typical_price, volume, period=14)
        assert len(result) == len(typical_price)
    
    def test_mfi_zero_period(self, test_data):
        """Test MFI fails with zero period - mirrors check_mfi_zero_period"""
        high = test_data['high'][:100]
        low = test_data['low'][:100]
        close = test_data['close'][:100]
        volume = test_data['volume'][:100]
        typical_price = (high + low + close) / 3.0
        
        with pytest.raises(ValueError, match="Invalid period"):
            ta_indicators.mfi(typical_price, volume, period=0)
    
    def test_mfi_period_exceeds_length(self):
        """Test MFI fails when period exceeds data length - mirrors check_mfi_period_exceeds_length"""
        input_high = np.array([1.0, 2.0, 3.0])
        input_low = np.array([0.5, 1.5, 2.5])
        input_close = np.array([0.8, 1.8, 2.8])
        input_volume = np.array([100.0, 200.0, 300.0])
        typical_price = (input_high + input_low + input_close) / 3.0
        
        with pytest.raises(ValueError, match="Invalid period"):
            ta_indicators.mfi(typical_price, input_volume, period=10)
    
    def test_mfi_very_small_dataset(self):
        """Test MFI fails with insufficient data - mirrors check_mfi_very_small_dataset"""
        input_high = np.array([1.0])
        input_low = np.array([0.5])
        input_close = np.array([0.8])
        input_volume = np.array([100.0])
        typical_price = (input_high[0] + input_low[0] + input_close[0]) / 3.0
        typical_price = np.array([typical_price])
        
        with pytest.raises(ValueError, match="Invalid period"):
            ta_indicators.mfi(typical_price, input_volume, period=14)
    
    def test_mfi_empty_input(self):
        """Test MFI fails with empty input"""
        empty = np.array([])
        
        with pytest.raises(ValueError, match="Empty data"):
            ta_indicators.mfi(empty, empty, period=14)
    
    def test_mfi_mismatched_lengths(self):
        """Test MFI fails with mismatched array lengths"""
        typical_price = np.array([1.0, 2.0, 3.0])
        volume = np.array([100.0, 200.0])
        
        
        
        with pytest.raises(ValueError, match="Empty data"):
            ta_indicators.mfi(typical_price, volume, period=2)
    
    def test_mfi_nan_handling(self, test_data):
        """Test MFI handles NaN values correctly"""
        high = test_data['high']
        low = test_data['low']
        close = test_data['close']
        volume = test_data['volume']
        typical_price = (high + low + close) / 3.0
        
        result = ta_indicators.mfi(typical_price, volume, period=14)
        assert len(result) == len(typical_price)
        
        
        assert np.all(np.isnan(result[:13])), "Expected NaN in warmup period"
        
        
        if len(result) > 13:
            
            valid_after_warmup = result[13:]
            assert not np.all(np.isnan(valid_after_warmup)), "Expected valid values after warmup"
            
            
            valid_values = valid_after_warmup[~np.isnan(valid_after_warmup)]
            assert np.all(valid_values >= 0.0), "MFI values should be >= 0"
            assert np.all(valid_values <= 100.0), "MFI values should be <= 100"
    
    def test_mfi_with_nan_input(self):
        """Test MFI with NaN values in input"""
        
        n = 30
        typical_price = np.arange(n, dtype=float)
        typical_price[5] = np.nan  
        volume = np.ones(n) * 1000
        
        result = ta_indicators.mfi(typical_price, volume, period=14)
        assert len(result) == len(typical_price)
        
        
        
        
        period = 14
        assert np.all(np.isnan(result[:period-1])), "Expected NaNs during warmup"
        assert np.all(np.isnan(result[period-1:])), "Expected NaNs after warmup due to seed NaN"
    
    def test_mfi_all_nan_input(self):
        """Test MFI with all NaN values"""
        n = 20
        typical_price = np.full(n, np.nan)
        volume = np.full(n, np.nan)
        
        with pytest.raises(ValueError, match="All values are NaN"):
            ta_indicators.mfi(typical_price, volume, period=14)
    
    def test_mfi_zero_volume(self):
        """Test MFI with zero volume - mirrors check_mfi_zero_volume"""
        
        n = 30
        typical_price = 100.0 + np.arange(n, dtype=float)
        volume = np.zeros(n)
        
        result = ta_indicators.mfi(typical_price, volume, period=14)
        
        
        valid_values = result[~np.isnan(result)]
        assert_close(valid_values, 0.0, rtol=1e-10, 
                    msg="MFI should be 0 with zero volume")
    
    def test_mfi_streaming(self, test_data):
        """Test MFI streaming matches batch calculation"""
        high = test_data['high'][:100]  
        low = test_data['low'][:100]
        close = test_data['close'][:100]
        volume = test_data['volume'][:100]
        typical_price = (high + low + close) / 3.0
        
        
        batch_result = ta_indicators.mfi(typical_price, volume, period=14)
        
        
        stream = ta_indicators.MfiStream(period=14)
        stream_values = []
        
        for i in range(len(typical_price)):
            result = stream.update(typical_price[i], volume[i])
            stream_values.append(result if result is not None else np.nan)
        
        stream_values = np.array(stream_values)
        
        
        assert len(batch_result) == len(stream_values)
        
        
        
        assert np.all(np.isnan(stream_values[:14])), "Expected NaN during streaming warmup"
        
        
        
        for i in range(14, len(batch_result)):
            if np.isnan(batch_result[i]) and np.isnan(stream_values[i]):
                continue
            assert_close(batch_result[i], stream_values[i], rtol=1e-9, atol=1e-9,
                        msg=f"MFI streaming mismatch at index {i}")
    
    def test_mfi_batch(self, test_data):
        """Test MFI batch processing - mirrors check_batch_default_row"""
        high = test_data['high']
        low = test_data['low']
        close = test_data['close']
        volume = test_data['volume']
        typical_price = (high + low + close) / 3.0
        
        result = ta_indicators.mfi_batch(
            typical_price,
            volume,
            period_range=(14, 14, 0)  
        )
        
        assert 'values' in result
        assert 'periods' in result
        
        
        assert result['values'].shape[0] == 1
        assert result['values'].shape[1] == len(typical_price)
        
        
        default_row = result['values'][0]
        expected = EXPECTED_OUTPUTS['mfi']['last_5_values']
        
        
        assert_close(
            default_row[-5:],
            expected,
            rtol=0.0,
            atol=1e-1,
            msg="MFI batch default row mismatch"
        )
    
    def test_mfi_batch_multiple_periods(self, test_data):
        """Test batch MFI with multiple periods"""
        high = test_data['high'][:100]  
        low = test_data['low'][:100]
        close = test_data['close'][:100]
        volume = test_data['volume'][:100]
        typical_price = (high + low + close) / 3.0
        
        
        result = ta_indicators.mfi_batch(
            typical_price,
            volume,
            period_range=(10, 20, 5)
        )
        
        
        assert result['values'].shape == (3, 100)
        assert np.array_equal(result['periods'], [10, 15, 20])
        
        
        single_result = ta_indicators.mfi(typical_price, volume, period=10)
        assert_close(result['values'][0], single_result, rtol=1e-10,
                    msg="Batch first row should match single calculation")
        
        
        for i, period in enumerate([10, 15, 20]):
            row = result['values'][i]
            
            assert np.all(np.isnan(row[:period-1])), f"Expected NaN warmup for period={period}"
            
            if period < 100:
                assert not np.all(np.isnan(row[period-1:])), f"Expected valid values after warmup for period={period}"
    
    def test_mfi_kernel_selection(self, test_data):
        """Test MFI with different kernel options"""
        high = test_data['high'][:100]
        low = test_data['low'][:100]
        close = test_data['close'][:100]
        volume = test_data['volume'][:100]
        typical_price = (high + low + close) / 3.0
        
        
        result_scalar = ta_indicators.mfi(typical_price, volume, period=14, kernel="scalar")
        result_auto = ta_indicators.mfi(typical_price, volume, period=14, kernel=None)
        
        
        assert_close(result_scalar, result_auto, rtol=1e-10,
                    msg="Kernel results should match")
    
    def test_mfi_different_periods(self, test_data):
        """Test MFI with various period values"""
        high = test_data['high'][:100]
        low = test_data['low'][:100]
        close = test_data['close'][:100]
        volume = test_data['volume'][:100]
        typical_price = (high + low + close) / 3.0
        
        periods = [7, 14, 21]
        results = []
        
        for period in periods:
            result = ta_indicators.mfi(typical_price, volume, period=period)
            assert len(result) == len(typical_price)
            
            assert np.all(np.isnan(result[:period-1])), f"Expected NaN warmup for period={period}"
            results.append(result)
        
        
        for i in range(len(periods)-1):
            valid_indices = ~(np.isnan(results[i]) | np.isnan(results[i+1]))
            if np.any(valid_indices):
                assert not np.allclose(results[i][valid_indices], results[i+1][valid_indices]),\
                    f"Results should differ for periods {periods[i]} and {periods[i+1]}"


if __name__ == '__main__':
    pytest.main([__file__, '-v'])
